### PR TITLE
Allow using `is null` checks to validate GetService calls

### LINKED ISSUE #113: VSSDK006 does not recognize `is null` check
#### Bug description

- Version of the vssdk-analyzers: 16.3.14
- Analyzer rule: VSSDK006
- Error (exception message, type, and callstack where applicable): 

>Warning	VSSDK006	Check whether the result of GetService calls is null.

#### To Reproduce
Steps or code to reproduce the behavior:
Sample code:

            var imageService = ServiceProvider.GlobalProvider.GetService(typeof(SVsImageService)) as IVsImageService2;
            if (imageService is null)
            {
                return default;
            }

#### Expected behavior
no error since I'm checking result for null

#### Additional context
This code works fine:

            var imageService = ServiceProvider.GlobalProvider.GetService(typeof(SVsImageService)) as IVsImageService2;
            if (imageService == null)
            {
                return default;
            }

### PR DESCRIPTION
**Pull request by jimmylewis**
Closes #113

### CODE REVIEW

File: src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs [hunk 1]
```diff
@@ -184,14 +184,22 @@ where symbol.Equals(symbolAccessed)
             private bool IsNonNullCheck(SyntaxNode node, ISymbol symbol, SyntaxNodeAnalysisContext context)
             {
                 bool IsSymbol(SyntaxNode n) => symbol.Equals(context.SemanticModel.GetSymbolInfo(n, context.CancellationToken).Symbol);
-
-                if (node is IfStatementSyntax ifStatement &&
-                    ifStatement.Condition.DescendantNodesAndSelf().OfType<BinaryExpressionSyntax>().Any(
-                        o => (o.OperatorToken.IsKind(SyntaxKind.EqualsEqualsToken) || o.OperatorToken.IsKind(SyntaxKind.ExclamationEqualsToken))
-                            && (o.Left.IsKind(SyntaxKind.NullLiteralExpression) || o.Right.IsKind(SyntaxKind.NullLiteralExpression))
-                            && (IsSymbol(o.Left) || IsSymbol(o.Right))))
+                bool IsEqualsOrExclamationEqualsCheck(BinaryExpressionSyntax o) => (o.OperatorToken.IsKind(SyntaxKind.EqualsEqualsToken) || o.OperatorToken.IsKind(SyntaxKind.ExclamationEqualsToken))
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs:187]: Can you try adding `static` modifiers to these local functions to ensure they don't pick up closures and add more allocations to analyzing?

File: src/Microsoft.VisualStudio.SDK.Analyzers.Tests/VSSDK006CheckServicesExistAnalyzerTests.cs [hunk 2]
```diff
@@ -614,6 +614,58 @@ class Test : Package {
...
+    protected override void Initialize() {
+        base.Initialize();
+        this.svc = this.GetService(typeof(SVsBuildManagerAccessor)) as IVsBuildManagerAccessor;
+    }
+
+    void Foo() {
+        if (svc is null) {
+            return;
+        }
+        svc.BeginDesignTimeBuild();
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.SDK.Analyzers.Tests/VSSDK006CheckServicesExistAnalyzerTests.cs:661]: Can you add a test (and support for, if necessary) the alternate syntax:

```cs
if (svc is object) {
   svc.BeginDesignTimeBuild();
}
```

also

```cs
if (svc is { }) {
   svc.BeginDesignTimeBuild();
}
```

or of course:

```cs
if (svc is IVsBuildManagerAccessor) {
   svc.BeginDesignTimeBuild();
}
```

File: src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs (hunk 1 above)
Comment (jimmylewis) [reply src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs:187]: static local functions is a C# 8 feature :(

Also, these can't be static, as they depend on `IsSymbol`, which has closures on `symbol` and `context`.  However, if I'm looking at the ildasm (and understand the compiler correctly), these all end up on the same object so it's the same number of allocations, but slightly larger to accommodate the new methods.

image

File: src/Microsoft.VisualStudio.SDK.Analyzers.Tests/VSSDK006CheckServicesExistAnalyzerTests.cs (hunk 2 above)
Comment (jimmylewis) [reply src/Microsoft.VisualStudio.SDK.Analyzers.Tests/VSSDK006CheckServicesExistAnalyzerTests.cs:661]: Done for `is object` and `is IVsBuildManagerAccessor`.
`is { }` is a C# 8 pattern match syntax.

File: src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs (hunk 1 above)
Comment (AArnott) [reply src/Microsoft.VisualStudio.SDK.Analyzers/VSSDK006CheckServicesExistAnalyzer.cs:187]: Oh! That's great that C# optimizes that.
We can certainly bump the lang version to 8 if that helped, but I'm satisfied with your answer.

**Pull request merged**